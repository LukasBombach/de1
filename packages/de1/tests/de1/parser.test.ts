import Parser from "../../src/parser";

type ParserMethod = "char" | "short" | "int" | "intSigned" | "sha";

type BufferMethods = {
  UInt8: ["writeUInt8", number];
  UInt16BE: ["writeUInt16BE", number];
  UInt32LE: ["writeUInt32LE", number];
  Int32LE: ["writeInt32LE", number];
};

type EachRecord = {
  fn: ParserMethod;
  type: keyof BufferMethods;
  value: number;
};

describe("parser", () => {
  const bufferMethods: BufferMethods = {
    UInt8: ["writeUInt8", 2],
    UInt16BE: ["writeUInt16BE", 4],
    UInt32LE: ["writeUInt32LE", 4],
    Int32LE: ["writeInt32LE", 4],
  };

  function getBuffer(type: keyof BufferMethods, value: number): Buffer {
    const [method, alloc] = bufferMethods[type];
    const buffer = Buffer.alloc(alloc);
    buffer[method](value, 0);
    return buffer;
  }

  test("works with an empty buffer", () => {
    const buffer = Buffer.from([0]);
    const parser = new Parser(buffer);
    expect(parser.vars()).toStrictEqual({});
  });

  test.each`
    fn             | type          | value
    ${"char"}      | ${"UInt8"}    | ${0x12}
    ${"short"}     | ${"UInt16BE"} | ${0x1234}
    ${"int"}       | ${"UInt32LE"} | ${0x12345678}
    ${"intSigned"} | ${"Int32LE"}  | ${0x12345678}
  `("$fn processes a $type", ({ fn, type, value }: EachRecord) => {
    const buffer = getBuffer(type, value);
    const process = jest.fn(v => v.toString());
    const parser = new Parser(buffer)[fn]("val") as Parser<{ val: number }>;
    const processor = new Parser(buffer)[fn]("val", process) as Parser<{
      val: string;
    }>;
    expect(parser.vars().val).toBe(value);
    expect(processor.vars().val).toBe(value.toString());
    expect(process).toHaveBeenCalledWith(value);
  });

  test("sha processes a UInt32LE", () => {
    const buffer = getBuffer("UInt32LE", 12345678);
    const parser = new Parser(buffer).sha("val") as Parser<{
      val: string;
    }>;
    expect(parser.vars().val).toBe("bc614e");
  });
});
