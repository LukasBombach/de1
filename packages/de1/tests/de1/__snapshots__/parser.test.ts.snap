// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser char processes a $type 1`] = `
Object {
  "value": 18,
}
`;

exports[`parser char processes a $type 2`] = `
Object {
  "val": "18",
}
`;

exports[`parser int processes a $type 1`] = `
Object {
  "value": 305419896,
}
`;

exports[`parser int processes a $type 2`] = `
Object {
  "val": "305419896",
}
`;

exports[`parser intSigned processes a $type 1`] = `
Object {
  "value": 305419896,
}
`;

exports[`parser intSigned processes a $type 2`] = `
Object {
  "val": "305419896",
}
`;

exports[`parser methods can be chained to create an object 1`] = `
Object {
  "char": 18,
  "int": 305419896,
  "intSigned": 305419896,
  "sha": "12345678",
  "short": 4660,
}
`;

exports[`parser sha processes a $type 1`] = `
Object {
  "value": "12345678",
}
`;

exports[`parser short processes a $type 1`] = `
Object {
  "value": 4660,
}
`;

exports[`parser short processes a $type 2`] = `
Object {
  "val": "4660",
}
`;

exports[`parser variables can be nested 1`] = `
Object {
  "one": Object {
    "a": 18,
    "b": 18,
  },
  "two": Object {
    "a": 18,
    "b": 18,
  },
}
`;
